### PR TITLE
some small fix

### PR DESCRIPTION
**Pull request by anjianshi**

### CODE REVIEW

**Comment by jantimon** [comment]
Please describe what you did and make sure that all tests pass.

File: index.js [hunk 1]
```diff
@@ -13,6 +13,10 @@ var LoaderTargetPlugin = require('webpack/lib/LoaderTargetPlugin');
 var LibraryTemplatePlugin = require('webpack/lib/LibraryTemplatePlugin');
 var SingleEntryPlugin = require('webpack/lib/SingleEntryPlugin');
 
+var compileQueue = {
```
Comment (jantimon) [inline index.js]: Remove the global

File: index.js [hunk 2]
```diff
@@ -148,6 +152,7 @@ HtmlWebpackPlugin.prototype.apply = function(compiler) {
           html: compilation.assets[self.options.filename],
           plugin: self
         }, callback);
+        return null;
```
Comment (jantimon) [inline index.js]: Why?

File: index.js [hunk 3]
```diff
@@ -188,26 +193,33 @@ HtmlWebpackPlugin.prototype.compileTemplate = function(template, outputFilename,
...
-
-        reject('Child compilation failed:\n' + errorDetails);
-      } else {
-        this.built = this.hash !== entries[0].hash;
-        this.hash = entries[0].hash;
-        resolve(childCompilation.assets[outputOptions.filename]);
-      }
-    }.bind(this));
-  }.bind(this));
+    compileQueue[self.options.template] = (compileQueue[self.options.template] || Promise.resolve())
```
Comment (jantimon) [inline index.js]: Why do we need the queue?

File: index.js [hunk 4]
```diff
@@ -216,7 +228,7 @@ HtmlWebpackPlugin.prototype.compileTemplate = function(template, outputFilename,
  */
 HtmlWebpackPlugin.prototype.evaluateCompilationResult = function(compilation, compilationResult) {
   if(!compilationResult) {
-    return Promise.reject('The child compilation didn\'t provide a result');
+    return Promise.reject(new Error('The child compilation didn\'t provide a result'));
```
Comment (jantimon) [inline index.js]: Not necessary to wrap it inside an error

File: index.js [hunk 5]
```diff
@@ -240,7 +252,7 @@ HtmlWebpackPlugin.prototype.evaluateCompilationResult = function(compilation, co
   }
   return typeof newSource === 'string' || typeof newSource === 'function' ?
     Promise.resolve(newSource) :
-    Promise.reject('The loader "' + this.options.template + '" didn\'t return html.');
+    Promise.reject(new Error('The loader "' + this.options.template + '" didn\'t return html.'));
```
Comment (jantimon) [inline index.js]: Not necessary to wrap it inside an error

File: index.js [hunk 6]
```diff
@@ -280,7 +292,7 @@ HtmlWebpackPlugin.prototype.executeTemplate = function(templateFunction, chunks,
 HtmlWebpackPlugin.prototype.postProcessHtml = function(html, assets) {
   var self = this;
   if (typeof html !== 'string') {
-    return Promise.reject('Expected html to be a string but got ' + JSON.stringify(html));
+    return Promise.reject(new Error('Expected html to be a string but got ' + JSON.stringify(html)));
```
Comment (jantimon) [inline index.js]: Not necessary to wrap it inside an error

**Comment by anjianshi** [comment]
All descriptions are in the commit message.
https://github.com/ampedandwired/html-webpack-plugin/pull/162/commits

**Comment by anjianshi** [comment]
@jantimon 

- The first commit fixed #130 , just ignore it now.
- In the second commit, I'm trying to fix a bug: template compile failed (got "HtmlWebpackPlugin Error: No source available") in webpack watch mode. The details was in the commit message.
- The third commit fixed a warning from bluebird: "Warning: a promise was rejected with a non-error"
   http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-rejected-with-a-non-error
  It cannot pass the test because the error message produced by case "shows an error when a template fails to load" was changed. May be I should update the test case?

**Comment by jantimon** [comment]
Oh okay sorry just read through it.

I have added a new test case which passes: https://github.com/ampedandwired/html-webpack-plugin/commit/3f16e3491a86722f2abff83402b94749cadd02b0

Right now I trie to investigate how we could test the recompilation.

**Comment by jantimon** [comment]
Okay the problem seems to be the "inParallel" this allows b to run before a in some cases.

**Comment by jantimon** [comment]
What do you think about https://github.com/ampedandwired/html-webpack-plugin/commit/62d7cfb36b0874e0aea2c6fc71b21e633e2fcd99 ?

**Comment by anjianshi** [comment]
very nice!  :)

**Comment by jantimon** [comment]
Cool :+1:

**Comment by jantimon** [comment]
Hey @anjianshi - just wanted to let you know that sokras subcache #179 solved "HtmlWebpackPlugin Error: No source available" and we can drop the promise queue

**Comment by anjianshi** [comment]
oh, thanks! happy to see this

**Pull request closed without merging**